File: src/main/java/com/naminhyeok/fantazzk/room/domain/DraftBoard.java
package com.naminhyeok.fantazzk.room.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class DraftBoard {
    private final List<String> teamLeaderIds;
    private final DraftOrderStrategy strategy;
    private final int picksPerTeam;

    public DraftBoard(List<String> teamLeaderIds, DraftOrderStrategy strategy, int picksPerTeam) {
        this.teamLeaderIds = List.copyOf(teamLeaderIds);
        this.strategy = strategy;
        this.picksPerTeam = picksPerTeam;
    }

    public List<String> getTeamLeaderIds() {
        return teamLeaderIds;
    }

    public DraftOrderStrategy getStrategy() {
        return strategy;
    }

    public int getPicksPerTeam() {
        return picksPerTeam;
    }

    public List<String> pickOrder() {
        List<String> reversed = new ArrayList<>(teamLeaderIds);
        Collections.reverse(reversed);
        List<String> order = new ArrayList<>();
        for (int round = 0; round < picksPerTeam; round++) {
            if (strategy == DraftOrderStrategy.SNAKE && round % 2 == 1) {
                order.addAll(reversed);
            } else {
                order.addAll(teamLeaderIds);
            }
        }
        return List.copyOf(order);
    }

    public String currentTeamLeader(int turnIndex) {
        require(turnIndex >= 0, "드래프트 턴은 0 이상이어야 합니다");
        List<String> order = pickOrder();
        checkState(turnIndex < order.size(), "드래프트가 이미 종료되었습니다");
        return order.get(turnIndex);
    }

    public void requireTurnOwner(int turnIndex, String teamLeaderId) {
        checkState(currentTeamLeader(turnIndex).equals(teamLeaderId), "현재 턴이 아닙니다");
    }

    public DraftPickSettlement settlePick(int turnIndex, int assignedCountAfterPick) {
        return new DraftPickSettlement(turnIndex + 1, assignedCountAfterPick >= totalPickCount());
    }

    private int totalPickCount() {
        return teamLeaderIds.size() * picksPerTeam;
    }

    private static void require(boolean condition, String message) {
        if (!condition) {
            throw new IllegalArgumentException(message);
        }
    }

    private static void checkState(boolean condition, String message) {
        if (!condition) {
            throw new IllegalStateException(message);
        }
    }
}

File: room/model/src/main/kotlin/com/naminhyeok/fantazzk/room/model/DraftBoard.kt
package com.naminhyeok.fantazzk.room.model

data class DraftBoard(
    val teamLeaderIds: List<String>,
    val strategy: DraftOrderStrategy,
    val picksPerTeam: Int,
) {
    fun pickOrder(): List<String> {
        val reversedOrder = teamLeaderIds.asReversed()
        return (0 until picksPerTeam).flatMap { round ->
            when (strategy) {
                DraftOrderStrategy.SNAKE -> if (round % 2 == 0) teamLeaderIds else reversedOrder
                DraftOrderStrategy.FIXED -> teamLeaderIds
            }
        }
    }

    fun currentTeamLeader(turnIndex: Int): String {
        require(turnIndex >= 0) { "드래프트 턴은 0 이상이어야 합니다" }
        val order = pickOrder()
        check(turnIndex < order.size) { "드래프트가 이미 종료되었습니다" }
        return order[turnIndex]
    }

    fun requireTurnOwner(
        turnIndex: Int,
        teamLeaderId: String,
    ) {
        check(currentTeamLeader(turnIndex) == teamLeaderId) { "현재 턴이 아닙니다" }
    }

    fun settlePick(
        turnIndex: Int,
        assignedCountAfterPick: Int,
    ): DraftPickSettlement =
        DraftPickSettlement(
            nextTurnIndex = turnIndex + 1,
            completed = assignedCountAfterPick >= totalPickCount(),
        )

    private fun totalPickCount(): Int = teamLeaderIds.size * picksPerTeam
}

data class DraftPickSettlement(
    val nextTurnIndex: Int,
    val completed: Boolean,
)
